arm/proofs/aes.ml

```diff
@@ -0,0 +1,393 @@
+(*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0 OR ISC OR MIT-0
+ *)
+
+needs "Library/words.ml";;
+needs "compute.ml";;
+
+(** TODO:define a recursive definition for word_join_list and its CONV *)
+
+let pp_print_num fmt tm =
+  let n = dest_numeral tm in
+  pp_print_string fmt (string_of_num_hex n) in
+install_user_printer("pp_print_num",pp_print_num);;
+
+let TEST thms =
+  let rw = Compute.bool_compset () in
+  (* avoid folding the branches of conditional expression 
+     before evaluating its condition *)
+  let _ = Compute.set_skip rw `COND: bool -> A -> A -> A` (Some 1) in
+  let _ = num_compute_add_convs rw in
+  let _ = word_compute_add_convs rw in
+  let _ = Compute.add_thms [EL_CONS;LET_END_DEF] rw in
+  let _ = Compute.add_thms thms rw in
+  Compute.WEAK_CBV_CONV rw;;
+
+let input = new_definition
+  `input:128 word = word 0xdee0b81ebfb441275d52119830aef1e5`;;
+
+(*
+bits(16*16*8) GF2 = (
+        /*       F E D C B A 9 8 7 6 5 4 3 2 1 0       */
+        /*F*/ 0x16bb54b00f2d99416842e6bf0d89a18c<127:0> :
+        /*E*/ 0xdf2855cee9871e9b948ed9691198f8e1<127:0> :
+        /*D*/ 0x9e1dc186b95735610ef6034866b53e70<127:0> :
+        /*C*/ 0x8a8bbd4b1f74dde8c6b4a61c2e2578ba<127:0> :
+        /*B*/ 0x08ae7a65eaf4566ca94ed58d6d37c8e7<127:0> :
+        /*A*/ 0x79e4959162acd3c25c2406490a3a32e0<127:0> :
+        /*9*/ 0xdb0b5ede14b8ee4688902a22dc4f8160<127:0> :
+        /*8*/ 0x73195d643d7ea7c41744975fec130ccd<127:0> :
+        /*7*/ 0xd2f3ff1021dab6bcf5389d928f40a351<127:0> :
+        /*6*/ 0xa89f3c507f02f94585334d43fbaaefd0<127:0> :
+        /*5*/ 0xcf584c4a39becb6a5bb1fc20ed00d153<127:0> :
+        /*4*/ 0x842fe329b3d63b52a05a6e1b1a2c8309<127:0> :
+        /*3*/ 0x75b227ebe28012079a059618c323c704<127:0> :
+        /*2*/ 0x1531d871f1e5a534ccf73f362693fdb7<127:0> :
+        /*1*/ 0xc072a49cafa2d4adf04759fa7dc982ca<127:0> :
+        /*0*/ 0x76abd7fe2b670130c56f6bf27b777c63<127:0>
+    );
+*)
+let GF2 = new_definition `GF2:((128)word) list =
+  [ word 0x16bb54b00f2d99416842e6bf0d89a18c
+  ; word 0xdf2855cee9871e9b948ed9691198f8e1
+  ; word 0x9e1dc186b95735610ef6034866b53e70
+  ; word 0x8a8bbd4b1f74dde8c6b4a61c2e2578ba
+  ; word 0x08ae7a65eaf4566ca94ed58d6d37c8e7
+  ; word 0x79e4959162acd3c25c2406490a3a32e0
+  ; word 0xdb0b5ede14b8ee4688902a22dc4f8160
+  ; word 0x73195d643d7ea7c41744975fec130ccd
+  ; word 0xd2f3ff1021dab6bcf5389d928f40a351
+  ; word 0xa89f3c507f02f94585334d43fbaaefd0
+  ; word 0xcf584c4a39becb6a5bb1fc20ed00d153
+  ; word 0x842fe329b3d63b52a05a6e1b1a2c8309
+  ; word 0x75b227ebe28012079a059618c323c704
+  ; word 0x1531d871f1e5a534ccf73f362693fdb7
+  ; word 0xc072a49cafa2d4adf04759fa7dc982ca
+  ; word 0x76abd7fe2b670130c56f6bf27b777c63
+  ]`;;
+
+TEST [GF2] `GF2`;;
```

Comment:
What about using `assert` to describe what its expected result would be?

Reply:
Same for other `TEST` uses below too. :)

Reply:
Or instead of using `assert` you can also manually check the output and raise an exception with `failwith "some concrete error message"` if this is easier to debug. :)

Reply:
This function only does evaluation of a HOL-Light term and does not check result. I changed its name to `EVAL`, which I think is more appropriate and moved it to a new file called `utils.ml`. I'm using it for debugging purposes during development, but not for adding assertions. I also commented the tests out because some of them are a bit slow and I don't want to make CI run longer. Let me know if you want me to remove those comments.
Also, I like the idea of having an ASSERT function. I could add one to the new file `utils.ml` if requested.

Reply:
I've further thought about it. I think I will remove these tests, and I will keep utils.ml for future development use. The reason I want to remove them is because running them would slow down the CI and they are already sort of tested in random instruction simulation. Leaving them in comments is a bad idea because it won't evolve with the code changes.